map_view/display-device-location-with-nmea-data-sources/src/main/java/com/esri/samples/display_device_location_with_nmea_data_sources/DisplayDeviceLocationWithNmeaDataSourcesController.java

```diff
@@ -0,0 +1,203 @@
+import com.esri.arcgisruntime.mapping.ArcGISMap;
+import com.esri.arcgisruntime.mapping.BasemapStyle;
+import com.esri.arcgisruntime.mapping.Viewpoint;
+import com.esri.arcgisruntime.mapping.view.LocationDisplay;
+import com.esri.arcgisruntime.mapping.view.MapView;
+
+public class DisplayDeviceLocationWithNmeaDataSourcesController {
+
+  @FXML private MapView mapView;
+  @FXML private Label satelliteCount;
+  @FXML private Label satelliteID;
+  @FXML private Button startButton;
+  @FXML private Button stopButton;
+  @FXML private Label systemInfo;
+
+  private int count = 0;
+  private NmeaLocationDataSource nmeaLocationDataSource;
+  private List<NmeaSatelliteInfo> nmeaSatelliteInfo;
+  private Timeline timeline;
+
+  public void initialize() {
+
+    try {
+
+      // authentication with an API key or named user is required to access basemaps and other location services
+      String yourAPIKey = System.getProperty("apiKey");
+      ArcGISRuntimeEnvironment.setApiKey(yourAPIKey);
+
+      // create a map with the navigation basemap style and set it to the map view
+      ArcGISMap map = new ArcGISMap(BasemapStyle.ARCGIS_NAVIGATION);
+      mapView.setMap(map);
+
+      // set a viewpoint on the map view centered on Redlands, California
+      mapView.setViewpoint(new Viewpoint(new Point(-117.191, 34.0306, SpatialReferences.getWgs84()), 100000));
+
+      // create a new NMEA location data source
+      nmeaLocationDataSource = new NmeaLocationDataSource(SpatialReferences.getWgs84());
+      // set the NMEA location data source onto the map view's location display
+      LocationDisplay locationDisplay = mapView.getLocationDisplay();
+      locationDisplay.setLocationDataSource(nmeaLocationDataSource);
+      locationDisplay.setAutoPanMode(LocationDisplay.AutoPanMode.RECENTER);
+      // disable map view interaction, the location display will automatically center on the mock device location
+      mapView.setEnableMousePan(false);
+      mapView.setEnableKeyboardNavigation(false);
+      stopButton.setDisable(true);
+
+    } catch (Exception e) {
+      // on any error, display the stack trace.
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * Initializes the location data source, reads the mock data NMEA sentences, and displays location updates from that file
+   * on the location display. Data is pushed to the data source using a time line to simulate live updates, as they would
+   * appear if using real-time data from a GPS dongle.
+   */
+  @FXML
+  private void start () {
+
+    // initialize the location data source and prepare to begin receiving location updates when data is pushed. As
+    // updates are received, they will be displayed on the map
+    nmeaLocationDataSource.startAsync();
+
+    // prepare the mock data NMEA sentences
+    File simulatedNmeaDataFile = new File(System.getProperty("data.dir"), "./samples-data/redlands/Redlands.nmea");
+    if (simulatedNmeaDataFile.exists()) {
+
+      try  {
```

Comment:
My IDE is saying that the resource should be closed in a finally block. I know you've got it being closed on line 121 - does it need to be closed here before the rest of the method? Or could `bufferedReader` be moved down into a finally? So that it gets closed regardless of any other Exceptions etc. Maybe doesn't matter in a Sample, but thought I'd raise!

Reply:
I'm not getting the same prompt, but I have tweaked the code a little: perhaps that tweak has negated the need for it?

Reply:
SonarLint still prompting this, but maybe one for @colinanderson to comment on.

Reply:
Maybe you can use a [try with resources here](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html). If not then it is good practice to put the closing in a finally block so you don't leak the buffered reader